src/Caching/StackExchangeRedis/src/RedisCache.cs

```diff
@@ -117,7 +146,25 @@ internal RedisCache(IOptions<RedisCacheOptions> optionsAccessor, ILogger logger)
 
         token.ThrowIfCancellationRequested();
 
-        return await GetAndRefreshAsync(key, getData: true, token: token).ConfigureAwait(false);
+        var expandedKey = _expandKey(key);
+
+        await ConnectAsync(token).ConfigureAwait(false);
+        Debug.Assert(_cache != null);
+
+        var value = await _cache.HashGetAsync(expandedKey, _getLabels)
+#if NET6_0_OR_GREATER
```

Comment:
Is it acceptable to only timeout for .NET core? Should we use the existing helpers for full framework?

Reply:
I don't think it's acceptable. Here's a Polyfill for WaitAsync that's usable on older frameworks:

```csharp

using System.Diagnostics.CodeAnalysis;

namespace System.Threading.Tasks;

/// <summary>
/// Provides a set of static methods for <see cref="Task"/>.
/// </summary>
[ExcludeFromCodeCoverage]
internal static class TaskExtensions
{
#pragma warning disable VSTHRD003 // Avoid awaiting foreign Tasks
    /// <summary>
    /// Gets a <see cref="Task{TResult}"/> that will complete when the <paramref name="task"/> completes or when the specified <paramref name="cancellationToken"/> has cancellation requested.
    /// </summary>
    /// <typeparam name="TResult">The type of the task result.</typeparam>
    /// <param name="task">The task to wait on for completion.</param>
    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for a cancellation request.</param>
    /// <returns>The <see cref="Task{TResult}"/> representing the asynchronous wait.</returns>
    public static Task<TResult> WaitAsync<TResult>(this Task<TResult> task, CancellationToken cancellationToken)
    {
        if (task.IsCompleted || (!cancellationToken.CanBeCanceled))
        {
            return task;
        }

        if (cancellationToken.IsCancellationRequested)
        {
            return Task.FromCanceled<TResult>(cancellationToken);
        }

        return WaitTaskAsync(task, cancellationToken);
    }

    private static async Task<TResult> WaitTaskAsync<TResult>(Task<TResult> task, CancellationToken cancellationToken)
    {
        var tcs = new TaskCompletionSource<TResult>(TaskCreationOptions.RunContinuationsAsynchronously);
        using (cancellationToken.Register(
                   static o => ((TaskCompletionSource<TResult>)o!).SetCanceled(), tcs, false))
        {
            var t = await Task.WhenAny(task, tcs.Task).ConfigureAwait(false);
#pragma warning disable VSTHRD103 // Call async methods when in an async method
            return t.GetAwaiter().GetResult();
#pragma warning restore VSTHRD103 // Call async methods when in an async method
        }
    }
}
```

Reply:
Can we lobby @mgravell to fix the SDK to add support for cancellation tokens? WaitAsync (built in or not) is suboptimal.

Reply:
Absolutely; now that I'm here and can find official time to do it: no problem! However, to be clear: redis doesn't have cancellation, so:

- if the command has not yet been written, fine - we can avoid sending it
- if it has been written, though; any server-side effects are out of our hands

We've discussed this many times but haven't found the time. Happy to fix that.

Reply:
Right, but anything that happens in the core library will be better than this 😄

Reply:
Everyone is fine removing the usage of `WaitAsync` for now and wait for Redis to handle the cancellation token?